src/libraries/Microsoft.PowerFx.Connectors/Execution/HttpFunctionInvoker.cs

```diff
@@ -156,37 +159,67 @@ private HttpContent GetBody(string referenceId, bool schemaLessBody, Dictionary<
             }
         }
 
-        public async Task<FormulaValue> DecodeResponseAsync(HttpResponseMessage response)
+        public async Task<FormulaValue> DecodeResponseAsync(HttpResponseMessage response, FormulaType returnType)
         {
             // $$$ Do we need to check response media type to confirm that the content is indeed json?
             var json = await response.Content.ReadAsStringAsync();
+            var statusCode = (int)response.StatusCode;
+            var message = statusCode < 300 ? string.Empty : $"Connector call failed ({response.StatusCode}), request {response.RequestMessage.RequestUri}, request headers {FormatHeaders(response.RequestMessage.Headers)}";
+
+            return statusCode switch
+            {                
+                // 3xx, 4xx, 4xx
+                int i when i >= 300 => FormulaValue.NewError(
+                    new ExpressionError() 
+                    { 
+                        Kind = ErrorKind.Network, 
+                        Severity = ErrorSeverity.Critical, 
+                        Message = message 
+                    },
+                    returnType),
+                
+                // 1xx, 2xx
+                _ => string.IsNullOrWhiteSpace(json) ? FormulaValue.NewBlank(_returnType) : FormulaValue.FromJson(json)
```

Comment:
>FormulaValue.FromJson(json) [](http://example.com/codeflow?start=92&length=27)

This is a separate pre-existing issue, but as a future PR, this should really be swagger driven (in the same way that serializers are) 
For example, if swagger specifies a date; FromJson on a string can't tell whether it's a date or string or guid.

Reply:
I moved the top comment